drv/vsc7448/src/spi.rs

```diff
@@ -0,0 +1,276 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use crate::VscError;
+use drv_spi_api::SpiDevice;
+use ringbuf::*;
+use vsc7448_pac::{types::RegisterAddress, Vsc7448};
+
+#[derive(Copy, Clone, PartialEq)]
+enum Trace {
+    None,
+    Read { addr: u32, value: u32 },
+    Write { addr: u32, value: u32 },
+}
+
+/// This indicates how many bytes we pad between (writing) the address bytes
+/// and (reading) data back, during SPI transactions to the VSC7448.  See
+/// section 5.5.2 for details.  1 padding byte should be good up to 6.5 MHz
+/// SPI clock.
+pub const SPI_PAD_BYTES: u8 = 1;
+
+// Flags to tune ringbuf output while developing
+const DEBUG_TRACE_SPI: u8 = 1 << 0;
+const DEBUG_MASK: u8 = 0;
+
+/// Writes the given value to the ringbuf if allowed by the global `DEBUG_MASK`
+macro_rules! ringbuf_entry_masked {
+    ($mask:ident, $value:expr) => {
+        if (DEBUG_MASK & $mask) != 0 {
+            ringbuf_entry!($value);
+        }
+    };
+}
+ringbuf!(Trace, 16, Trace::None);
+
+////////////////////////////////////////////////////////////////////////////////
+
+/// Helper struct to read and write from the VSC7448 over SPI
+pub struct Vsc7448Spi(pub SpiDevice);
+impl Vsc7448Spi {
+    /// Reads from a VSC7448 register
+    pub fn read<T>(&self, reg: RegisterAddress<T>) -> Result<T, VscError>
+    where
+        T: From<u32>,
+    {
+        assert!(reg.addr >= 0x71000000);
```

Comment:
Having these preconditions mentioned in the comment on this function would be good.

(Text size nit: doing one assert using `&&` would likely be cheaper.)

Reply:
I switched to returning an error if the caller is out of range, instead of panicking.

Reply:
...and change my mind and switched back, after documenting the precondition.